File: Java/DP/MaxSumWithoutAdjacentElem.java
public class Solution {
	public int adjacent(ArrayList<ArrayList<Integer>> a) {
	    int size = a.get(0).size();
	    for(int i = 0; i < size; i++)
	        a.get(0).set(i, Math.max(a.get(0).get(i), a.get(1).get(i)));
	   // Now we'll use second row for memoization
	   ArrayList<Integer> A = a.get(0), DP = a.get(1);
	   if(size == 1)return A.get(0);
	   if(size < 3)return Math.max(A.get(0), A.get(1));

	   DP.set(0, A.get(0));
	   DP.set(1, Math.max(DP.get(0), A.get(1)));

	   for(int i = 2; i < size; i++){
	       DP.set(i, Math.max(DP.get(i-1), DP.get(i-2) + A.get(i)));
	   }
	   return DP.get(size-1);
	}
}

File: JS/DP/MaxSumWithoutAdjacentElem.js
module.exports = {
	//param A : array of array of integers
	//return an integer
	adjacent : function(A){
        var len = A[0].length;
        for(var i = 0; i < len; i++)
            A[0][i] = Math.max(A[0][i], A[1][i]);
        var aa = A[0], mem = A[1];
        mem[0] = aa[0];
        mem[1] = Math.max(aa[0], aa[1]);
        for(var i = 2; i < len; i++)
            mem[i] = Math.max(aa[i] + mem[i-2], mem[i-1]);
        return mem[len-1];
	}
};
